distributed/worker.py

```diff
@@ -1307,23 +1306,31 @@ def keys(self) -> list[str]:
         return list(self.data)
 
     async def gather(self, who_has: dict[str, list[str]]) -> dict[str, Any]:
+        """Endpoint used by Scheduler.replicate()"""
+        first = True
         who_has = {
             k: [coerce_to_address(addr) for addr in v]
             for k, v in who_has.items()
             if k not in self.data
         }
-        result, missing_keys, missing_workers = await gather_from_workers(
-            who_has=who_has, rpc=self.rpc, who=self.address
+
+        async def refresh_who_has(keys: list[str]) -> Mapping[str, Collection[str]]:
+            nonlocal first
+            if first:
+                first = False
```

Comment:
why this? Isn't the point of this callback architecture to be able to update this on every iteration?

Reply:
This avoids introducing an unnecessary additional round-trip from worker to scheduler every time the user calls replicate().
If you think that code cleanliness trumps performance in this very niche use case, I'm happy to simplify the logic.